internal/testsuite/eventwatcher/eventwatcher.go

```diff
@@ -0,0 +1,266 @@
+	numActive := 0
+	numRemaining := len(jobIds)
+	exitedByJobId := make(map[string]bool)
+	for {
+		select {
+		case <-parent.Done():
+			return nil
+		case msg := <-C:
+			if e := msg.GetSubmitted(); e != nil {
+				numActive++
+			} else if e := msg.GetSucceeded(); e != nil {
+				if _, ok := exitedByJobId[e.JobId]; ok {
+					return fmt.Errorf("received multiple terminal events for job %s", e.JobId)
+				}
+				exitedByJobId[e.JobId] = true
+				if _, ok := jobIds[e.JobId]; ok {
+					numRemaining--
+				}
+				numActive--
+			} else if e := msg.GetFailed(); e != nil {
+				if _, ok := exitedByJobId[e.JobId]; ok {
+					return fmt.Errorf("received multiple terminal events for job %s", e.JobId)
+				}
+				exitedByJobId[e.JobId] = true
+				if _, ok := jobIds[e.JobId]; ok {
+					numRemaining--
+				}
+				numActive--
+			}
+			if numRemaining <= 0 {
+				return errors.New("all jobs exited")
+			}
+		}
+	}
+}
+
+// WithCancelOnFailed returns a context that is cancelled if a job fails,
+// or if the parent is cancelled.
+func ErrorOnFailed(parent context.Context, C chan *api.EventMessage) error {
+	for {
+		select {
+		case <-parent.Done():
+			return parent.Err()
+		case msg := <-C:
+			if failedEvent := msg.GetFailed(); failedEvent != nil {
+				err := fmt.Errorf("job failed")
+				return errors.WithMessagef(err, "%v", failedEvent)
+			}
+		}
+	}
+}
+
+// GetFromIngresses listens for ingressInfo messages and tries to download from each ingress.
+// Returns false if any download fails.
+func GetFromIngresses(parent context.Context, C chan *api.EventMessage) error {
+	g, ctx := errgroup.WithContext(parent)
+	for {
+		select {
+		case <-parent.Done():
+			return g.Wait()
+		case <-ctx.Done(): // errgroup cancelled
+			return g.Wait()
+		case msg := <-C:
+			if ingressInfo := msg.GetIngressInfo(); ingressInfo != nil {
+				for _, host := range ingressInfo.IngressAddresses {
+					ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+					defer cancel()
+					host := host
+					g.Go(func() error { return getFromIngress(ctx, host) })
+				}
+			}
+		}
+	}
+}
+
+func getFromIngress(ctx context.Context, host string) error {
+	tokens := strings.Split(host, ".")
+	if len(tokens) == 0 {
+		return fmt.Errorf("malformed hostname: %s", host)
+	}
+	url := "http://" + tokens[len(tokens)-1]
+
+	// The ingress info messages can't convey which port ingress are handled on (only the url).
+	// (The assumption is that ingress is always handled on port 80.)
+	// Here, we override this with an environment varibel so we can test against local Kind clusters,
+	// which may handle ingress on an arbitrary port.
+	ingressPort := os.Getenv("ARMADA_EXECUTOR_INGRESS_PORT")
+	if ingressPort == "" {
+		ingressPort = "80"
+	}
+
+	// Make a get request to test that the ingress works.
+	// This assumes that whatever the ingress points to responds.
+	httpClient := &http.Client{}
+	httpReq, err := http.NewRequest("GET", url+":"+ingressPort+"/", nil)
```

Comment:
I'd rather use `NewRequestWithContext` instead of `NewRequest`.

For the second param, my preference is always in favour of `fmt.Sprintf`

i.e. `fmt.Sprintf("%s:%s/", url, ingressPort).

For the body param, a lot of people suggest to use `http.NoBody` in favour to `nil`

Reply:
Done